### PR TITLE
Feature/pil 153

### PR DESCRIPTION
**Pull request by ricardohg**
this is a draft for Date of age patch, test are still missing but business logic should be completed, please let me know if you see something weird or missing.

### CODE REVIEW

File: Sources/Apollo/GraphQLFirstReceivedAtTracker.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  func accept(list: [Date], info: FieldExecutionInfo) throws -> Date {
+    return list.min() ?? .distantPast
+  }
+
+  func accept(fieldEntry: Date, info: FieldExecutionInfo) throws -> Date? {
+    return fieldEntry
+  }
+
+  func accept(fieldEntries: [Date], info: ObjectExecutionInfo) throws -> Date {
+    return fieldEntries.min() ?? .distantPast
```
Comment (Mordil) [inline Sources/Apollo/GraphQLFirstReceivedAtTracker.swift]: ❗ You'll want to make sure you're not making the same bug mistakes I made https://github.com/peek-travel/apollo-ios/pull/7/files

By doing `min` we end up with the oldest date which always ends up being `distantPast`

File: Sources/Apollo/GraphQLResult.swift [hunk 2]
```diff
@@ -3,6 +3,28 @@ import Foundation
...
+
+extension GraphQLResultMetadata: Hashable {
+
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(self.id)
+  }
+
+  public static func == (lhs: GraphQLResultMetadata, rhs: GraphQLResultMetadata) -> Bool {
+   return lhs.maxAge == rhs.maxAge
+  }
```
Comment (Mordil) [inline Sources/Apollo/GraphQLResult.swift]: ❗ This violates the semantics of hashable & equatable.

You should be using the same properties in both.

File: Sources/Apollo/GraphQLResultAccumulator.swift [hunk 3]
```diff
@@ -8,9 +8,9 @@ protocol GraphQLResultAccumulator: AnyObject {
   associatedtype ObjectResult
   associatedtype FinalResult
 
-  func accept(scalar: JSONValue, info: FieldExecutionInfo) throws -> PartialResult
-  func acceptNullValue(info: FieldExecutionInfo) throws -> PartialResult
-  func acceptMissingValue(info: FieldExecutionInfo) throws -> PartialResult
+  func accept(scalar: JSONValue, firstReceivedAt: Date, info: FieldExecutionInfo) throws -> PartialResult
+  func acceptNullValue(firstReceivedAt: Date, info: FieldExecutionInfo) throws -> PartialResult
+  func acceptMissingValue(firstReceivedAt: Date, info: FieldExecutionInfo) throws -> PartialResult
```
Comment (Mordil) [inline Sources/Apollo/GraphQLResultAccumulator.swift:13]: ❓ I don't remember why I marked them as `Date?` in the original changes - is that not necessary this time?

**Review by Mordil — CHANGES_REQUESTED** [review]

File: Sources/Apollo/GraphQLResultAccumulator.swift (hunk 3 above)
Comment (ricardohg) [reply Sources/Apollo/GraphQLResultAccumulator.swift:13]: if you are talking about `firstReceivedAt` parameter, for what I see it was never `Date?`

File: Sources/Apollo/GraphQLResult.swift (hunk 2 above)
Comment (ricardohg) [reply Sources/Apollo/GraphQLResult.swift]: done

File: Sources/Apollo/GraphQLFirstReceivedAtTracker.swift (hunk 1 above)
Comment (ricardohg) [reply Sources/Apollo/GraphQLFirstReceivedAtTracker.swift]: added

File: Sources/Apollo/GraphQLResultAccumulator.swift (hunk 3 above)
Comment (Mordil) [reply Sources/Apollo/GraphQLResultAccumulator.swift:13]: If you look at the PR that fixed the min vs max age that I linked in another comment - I changed the values to `Date?`

I wonder if it was me refining the protocol, and giving a quirk of how protocol conformance works

Comment (ricardohg) [reply Sources/Apollo/GraphQLResultAccumulator.swift:13]: I think it was you refining the protocol

Comment (Mordil) [reply Sources/Apollo/GraphQLResultAccumulator.swift:13]: Got it, never mind then

**Comment by ricardohg** [comment]
@Mordil @filipealva this is now ready for review!

File: Tests/ApolloInternalTestHelpers/SQLiteTestCacheProvider.swift [hunk 4]
```diff
@@ -5,12 +5,9 @@ import ApolloSQLite
...
   /// Execute a test block rather than return a cache synchronously, since cache setup may be
   /// asynchronous at some point.
-  public static func withCache(initialRecords: RecordSet? = nil, fileURL: URL? = nil, execute test: (NormalizedCache) throws -> ()) throws {
+  public static func withCache(initialRecords: RecordSet? = nil, fileURL: URL? = nil, execute test: (NormalizedCache) throws -> ()) rethrows {
     let fileURL = fileURL ?? temporarySQLiteFileURL()
-    let cache = try! SQLiteNormalizedCache(fileURL: fileURL)
-    if let initialRecords = initialRecords {
-      _ = try cache.merge(records: initialRecords)
-    }
+    let cache = try! SQLiteNormalizedCache(fileURL: fileURL, initialRecords: initialRecords)
```
Comment (Mordil) [inline Tests/ApolloInternalTestHelpers/SQLiteTestCacheProvider.swift:10]: ❔ Was this in the original commit? 🤔 We shouldn't force unwrap if we're rethrowing IMO

but the original code has it... so I guess it's moot

**Review by Mordil — COMMENTED** [review]
Just for sanity sake, could you post a screenshot/gif of all the unit tests passing?

**Comment by ricardohg** [comment]
Screenshot 2023-02-03 at 2 50 23 PM

**Comment by ricardohg** [comment]
> Just for sanity sake, could you post a screenshot/gif of all the unit tests passing?

done

**Review by Mordil — APPROVED** [review]

**Pull request merged**